### DISCUSSION TITLE
How can i pass payload to http.put?

### DISCUSSION THREAD

**Discussion by rrolla** [savonet/liquidsoap#2068] [General]
Im trying to do simple put where it updates song bpm on db, but when i pass data argument put request didn't reach server

OS: ubuntu:20.04 on docker
Liquidsoap version: https://github.com/savonet/liquidsoap/releases/tag/v2.0.0 installed from deb package

### This doesn't reach server
```ruby
#!/usr/bin/liquidsoap

mainSource = single("fallback.mp3")
mainSource = bpm(mainSource)

def detectBpm()
  bpm = mainSource.bpm()
  print("BPM: " ^ string_of(bpm))
  url = "http://utilsnginx/api/tunes/bpm"
  print("put request to: " ^ url)
  j = json()
  j.add("bpm", bpm)
  putData=json.stringify(j)
  print(putData)
  http.put(headers=[("Content-Type","application/json"),("Accept","application/json")], url, data=putData)
  {}
end

detectBpm()

output.dummy(fallible=true, mainSource)
```

output from liquidsoap
```
rolla-liquidsoap | BPM: 0.
rolla-liquidsoap | put request to: http://utilsnginx/api/tunes/bpm
rolla-liquidsoap | { "bpm": 0.0
rolla-liquidsoap | }
```

### Without passing data it reaches server and also executes much faster
```ruby
#!/usr/bin/liquidsoap

mainSource = single("fallback.mp3")
mainSource = bpm(mainSource)

def detectBpm()
  bpm = mainSource.bpm()
  print("BPM: " ^ string_of(bpm))
  url = "http://utilsnginx/api/tunes/bpm"
  print("put request to: " ^ url)
  http.put(headers=[("Content-Type","application/json"),("Accept","application/json")], url)
  {}
end

detectBpm()

output.dummy(fallible=true, mainSource)
```
output from liquidsoap
```
rolla-liquidsoap | BPM: 0.
rolla-liquidsoap | put request to: http://utilsnginx/api/tunes/bpm
```

![image](https://user-images.githubusercontent.com/3932546/142749223-ee10e3d7-4d48-43ec-8232-59adbec649bc.png)

Any thoughts?

**Comment by smimram** [comment]
Maybe is it the newline in the JSON? Can you try with `json.stringify(compact=true, ...)`?

**Reply by rrolla** [reply]
@smimram Unfortunately, it also doesn't work if you pass compact=true. Also if i'm removing headers it doesn't reach server.

Can you please post simple PUT example with payload that works?

Just tried http.post and that works, i will use that one.

**Reply by toots** [reply]
I just added a test for `put` and noticed that the difference between post and put is that `libcurl` uses the `expect-100` request header:
```
   test.equals(req.headers, [
      ("host", "localhost:5427"),
      ("user-agent", http.user_agent),
      ("transfer-encoding", "chunked"),
      ("content-type", "application/json"),
      ("accept", "application/json"),
      ("expect", "100-continue")
    ])
```
You should make sure that your server recognizes and knows to response with the `100` status otherwise the `PUT` query will indeed be stuck.

**Reply by rrolla** [reply]
Thanks, @toots for finding the root cause.
